akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala

```diff
@@ -147,14 +143,27 @@ private[projection] object DynamoDBOffsetStore {
       }
     }
 
-    def evict(slice: Int, timeWindow: JDuration): State = {
+    def evict(slice: Int, timeWindow: JDuration, ableToEvictRecord: Record => Boolean): State = {
       val recordsSortedByTimestamp = bySliceSorted.getOrElse(slice, TreeSet.empty[Record])
       if (recordsSortedByTimestamp.isEmpty) {
         this
       } else {
         val until = recordsSortedByTimestamp.last.timestamp.minus(timeWindow)
-        val filtered = recordsSortedByTimestamp.dropWhile(_.timestamp.isBefore(until))
-        if (filtered.size == recordsSortedByTimestamp.size) {
+        val filtered = {
+          // Records comparing >= this record by recordOrdering will definitely be kept,
+          // Records comparing < this record by recordOrdering are subject to eviction
+          // Slice will be equal, and pid will compare lexicographically less than any valid pid
+          val untilRecord = Record(slice, "", 0, until)
+          val newerRecords = recordsSortedByTimestamp.rangeImpl(Some(untilRecord), None) // inclusive of until
+          val olderRecords = recordsSortedByTimestamp.rangeImpl(None, Some(untilRecord)) // exclusive of until
```

Comment:
I think clearer to call `rangeFrom`/`rangeUntil` rather than the underlying `rangeImpl`:

```suggestion
          val newerRecords = recordsSortedByTimestamp.rangeFrom(untilRecord) // inclusive of until
          val olderRecords = recordsSortedByTimestamp.rangeUntil(untilRecord) // exclusive of until
```